bindings/serialize-v8.cc

```diff
@@ -0,0 +1,205 @@
+ * limitations under the License.
+ */
+
+#include "serialize-v8.h"
+#include "serialize.h"
+
+using namespace v8;
+
+class HeapNode : public Node {
+ private:
+  AllocationProfile::Node *node;
+
+ public:
+  HeapNode(AllocationProfile::Node *node) { this->node = node; }
+
+  virtual std::string name() { return *String::Utf8Value(node->name); }
+
+  virtual std::string filename() {
+    return *String::Utf8Value(node->script_name);
+  }
+
+  virtual int64_t getFileID() { return node->script_id; }
+
+  virtual int64_t lineNumber() { return node->line_number; }
+
+  virtual int64_t columnNumber() { return node->column_number; }
+
+  virtual std::vector<Sample> samples(const std::deque<uint64_t> &stack,
+                                      Profile *p) {
+    std::vector<Sample> samples;
+    for (size_t i = 0; i < node->allocations.size(); i++) {
+      AllocationProfile::Allocation allocation = node->allocations[i];
+      std::vector<Label> labels = {Label(
+          p->stringID("allocation"), 0, allocation.size, p->stringID("bytes"))};
+      std::vector<int64_t> vals = {
+          allocation.count,
+          static_cast<int64_t>(allocation.size * allocation.count)};
+      Sample s = Sample({stack.begin(), stack.end()}, vals, labels);
+      samples.push_back(s);
+    }
+    return samples;
+  }
+};
+
+class TimeNode : public Node {
+ private:
+  const CpuProfileNode *node;
+  int samplingIntervalMicros;
+
+ public:
+  TimeNode(const CpuProfileNode *node, int samplingIntervalMicros) {
+    this->node = node;
+    this->samplingIntervalMicros = samplingIntervalMicros;
+  }
+
+  virtual std::string name() {
+    return *String::Utf8Value(node->GetFunctionName());
+  }
+
+  virtual std::string filename() {
+    return *String::Utf8Value(node->GetScriptResourceName());
+  }
+
+  virtual int64_t getFileID() { return node->GetScriptId(); }
+
+  virtual int64_t lineNumber() { return node->GetLineNumber(); }
+
+  virtual int64_t columnNumber() { return node->GetColumnNumber(); }
+
+  virtual std::vector<Sample> samples(const std::deque<uint64_t> &stack,
+                                      Profile *p) {
+    std::vector<Sample> samples;
+    int64_t hitCount = node->GetHitCount();
+    std::vector<int64_t> vals = {hitCount, hitCount * samplingIntervalMicros};
+    Sample s = Sample({stack.begin(), stack.end()}, vals, std::vector<Label>());
+    samples.push_back(s);
+    return samples;
+  }
+};
+
+struct TimeEntry {
+  const CpuProfileNode *node;
+
+  // number of entries which would need to be removed from the stack after
+  // processing the node if the node were a leaf node.
+  int popCount;
+};
+
+std::unique_ptr<std::vector<char>> serializeTimeProfile(
+    CpuProfile *profile, int64_t samplingIntervalMicros,
+    int64_t startTimeNanos) {
+  int64_t durationNanos =
+      (profile->GetEndTime() - profile->GetStartTime()) * 1000;
+
+  Profile p = Profile("wall", "microseconds", samplingIntervalMicros,
```

Comment:
`profile` would be a more descriptive name than `p`.

Reply:
done.